### ISSUE TITLE
Read-only mode

### ISSUE THREAD

**Issue by lastzero** [photoprism/photoprism#56]
**As a user with a well-managed photo collection, I want to run the application in a mode that doesn't touch my directory or file names in order to keep the existing structure.**

For those who want to manage their files manually or with other photo management tools, see #41. We also need this for a public demo. See also [Concerns](https://github.com/photoprism/photoprism/wiki/Concerns).

**Comment by lastzero** [comment]
Maybe we can use this issue to build a proof-of-concept for feature flags and general settings. Not everything should be configured in the main config file. Need to think about it.

**Comment by dmag** [comment]
I would say this is a _critical_ issue for this software. Your target market seems to be "people who are paranoid enough to keep their photos out of the cloud", but those same people will be paranoid about giving random software "write access" to their photos.

So to try this software, they must manually-copy over a small subset of their photos. (Not everyone has enough space to duplicate their photo collection.)  That small subset makes it harder to see the full power of this software.

For photo collections that don't have a database, I can see the need to manage images via the filesystem. But if the software requires a database, then the physical filename **does not** matter. Keeping photos Read-Only is also "good security". (See article "Man Sues Adobe Over Bug That Deleted His Work".)

Looking forward to the MVP!

**Comment by lastzero** [comment]
@dmag You don't have to copy your photos, you can simply set your originals path to whatever directory you use.

Specific issues are:
- Our indexer doesn't expect duplicates in the originals directory (yet), so images will only be indexed once. Doesn't really matter, as long as you don't want to delete or move any images(s). A read-only mode can prevent that.
- Keeping track of duplicates is possible, but clutters the database and makes the software more complex. That's why we suggest using the Import -> Originals workflow where duplicates get sorted out and all files get a canonical name. It only depends on the metadata and file hash, so you don't even need a database lookup or connection.
-  When you edit one of the duplicates, you might have to implement forking in your database. Or you say file names don't matter at all, but hashes can change when users edit images. That means photos that were in an album can suddenly disappear. Of course you can think about all this very well, but it's not fun and makes our work harder than needed.
- If you want the use our software to add / upload files, it must know how to name them. Doesn't matter if your are paranoid. You can not have it all.

**Comment by lastzero** [comment]
Having said that, we welcome reports of users who try to index their existing photo library (please experiment with a copy). If you have RAWs, PhotoPrism might try to create JPEGs from them in order to show thumbnails etc. It is possible that it uses the common name (our naming scheme based on metadata) for that instead of [name of the RAW].jpg, but that could be solved easily. I'm going to have a look myself after we have complete test coverage. One step after another.

**Comment by matthijsvdr** [comment]
So I just came across of photoprism and it looks like the thing I have been looking for (started to think I would have to build something my self). 
I was reading the project status and ended up here, just wanted to give me 2 cents on how I mange my photos (maybe if gives you some idea's or you can just ignore it 😄).

So currently I have a "single source of truth". Got a Synology NAS where all photos are organised in folders (Country -> Year -> City/region -> Activity/ or something like that and work related Year -> 
Client -> Types of work)

So Synology has 2 photo solutions:
**Moments**
You have a timeline view for all photos, but you can also switch to folder view (how ever the folder view is just that, and you are unable to set cover images/ share them or anything else).
Create albums and add photos from anywhere on disk or upload them.
You have some ML going on for tagging and face detection etc

**Photostation** 
Each folder is an album, and you could apply share settings/album image/premissions etc on each one.
Does not have any ML


So I like Photostation because it allows me to just create folders on disk, and it will create albums of it. Then I can easily browse them on a web page, and share them with family fiends etc.
But I like Moments because of the timeline view and the ML tagging and the ability to create albums not linked the the folder structure.

In a perfect world these 2 would be combined. So each folders is also an album and all options that come with it (allowing me to manage things on disk, as I connect over SMB to the nas when I edit or sync things when I'm remote etc) / Creating extra albums that are not related to folder structures and have all the great features of ML that will help me find things without manually tagging images.

Greetings,
Matthijs

**Comment by lastzero** [comment]
Thanks for the feedback! We do our best to get this done... note that this is our private time & money, so if you want special features, consider a donation or send a pull request 👍 

PhotoPrism will allow you to edit files on the disk, but directories are not albums by default... although I assume some users want to manage and share their albums with Windows Explorer. They might also create ZIP files of their 50 MP full resolution JPEGs (high compression to save storage) and send them via email 🥳

Our primary target group are users with too many photos and high quality standards. Like us. We do use Photoshop and edit files on the disk, but we don't want full control over directories and file names because it's just too much. You also don't manage single files and directories when you upload your images to google photos, flickr or facebook. So I guess it will be something in between.

**Comment by friesoft** [comment]
Another Synology user here using Photostation. We (as in my family and friends) also really like to organize our photos in folder structures like this:
- 2018
-- 2018-01-01_New_Year
- 2019
-- 2019-06-19_Awesome_Event
-- 2019-06-20_Some_Event

This works great for us but leaves us with the crippled Photostation with its abysmal permission system (who in his right mind automatically assigns every user permission to new folders???)

We are searching for an alternative to Photostation and your project sounds great with some areas worrying me a bit:
* no read only mode (this issue) => our photo collection is > 1TB - no I don't let anything modify its structure automatically
* folders should be automatically imported as albums (and be kept in sync!)
* all changes on the filesystem should be reflected in the albums (renames, moves, ...)
* I want to see my folder structure reflected as albums (fancy views like sort by date, ... are nice but only the cherry on top - If I carefully arrange photos together in a folder there's are reason for it...

I guess all my fellow synology users are looking for photostation with a better interface, better permission system and the cool stuff like face recognition, search, ...

Basically the only projects I found lately coming close to what we are searching for at the moment are photoprism and ownphotos.

Do you think our way of doing stuff will be something supported by photoprism in the future?

PS: I guess I qualify with this statement and > 1TB of photos :)
> Our primary target group are users with too many photos

**Comment by lastzero** [comment]
@friesoft Thank you for the feedback! It's really valuable for us but it also makes me think if we should publish a "Pro" version for like $ 25 / year. I don't see us working full-time on features we don't need for ourselves otherwise while our users are getting paid for their jobs and enjoy the software for free.

**Comment by friesoft** [comment]
I would definitely be willing to pay such a price for a well working software I can trust my photos. I would prefer though if also the pro features where opensource so everyone can improve those features together. A model like elastic is doing with the xpack would be cool.

**Comment by lastzero** [comment]
@friesoft Another option would be to keep the source free and sell an optimized / tested binary like the Caddy team does: https://caddyserver.com/download

Also we started working on an appliance, see Twitter: https://twitter.com/browseyourlife

Funding our project with donations would be great but even GitLab never received more than $ 1000 / month. That's obviously not enough for a team to work full-time.

**Comment by FrouxBY** [comment]
After having a look to photoprism as well and have a little experiences with synology, I will share my thoughts on it,

Definitely the combo with PhotoStation which handle library as the folder structures is and Moments from Synology are something very interesting. However Moments is intended to be more a Google Photos apps replacement, and I think Photoprism as well.

> some areas worrying me a bit:
> no read only mode (this issue) => our photo collection is > 1TB - no I don't let anything modify its structure automatically

I think this point is really important, as it will definitely be a **vector of adoption by users**. Indeed, most people with large databases of photo won't trust any service to modify too much their file/structure, and have a chance to mess up with it.

> I want to see my folder structure reflected as albums (fancy views like sort by date, ... are nice but only the cherry on top - If I carefully arrange photos together in a folder there's are reason for it...

For this point, I disagree a little. The big interest of Moments or Photoprism is to have 'smart' albums, which are fully independant of the file structure. This make possible to have any arrangement of your photos without dealing with copying it arround, synchronising it altogether at each change, etc. If you really want that all albums are reflected in files structures, then PhotoStation or Nextcloud Gallery will do it quite well. But for smarter management, you need too rely on a database and not on file structure.
I would add that this make possible to keep only one copy of each file, and this is really important to keep a photo collection as clean as possible.

However, the other way arround seems more reasonnable : having file structure reflected as album. 
This can be done easily for the first import, a kind of migration process.  It would be even better if this is done without changing the directory. There would be some open questions, such as "folders in folders reflect as albums in albums ?" but it will be handable easily? However, what after this first big transition ? How to watch for folder change, etc ? This is much more complicated to implement and maybe not necessary, I will explain what can be a work arround next.

This is my vision of how it could work **If Photoprism is intended to manage files on its own**. First an import process on place which track folder structure to keep already created albums in order to create them in Photoprism. After this initial import, Photoprism would have the management of those folders exclusively (meaning that external changes require to manually trigger a import/index process), and user will add new pictures and manage them and albums **through Photoprism only**. Finally if the user want to get out Photoprism and using whatever other photo management software, Photoprism may have a kind of _export_ fonction **in place** which give back the library as a file structure and not only one big folder with all the photos in it and no meaningful name.

This export function is really not difficult to provide, most people are using date folder structure, and datetime as filename are a smart way to rename it. Photoprism do even better with a hash avoiding collision with different camera taking picture at the exact same time. This mean than folder structure created are already quite usable in their own, except that date only loose the title of the album. So what I would do is adding a possibility to **associate event with range of datetime**. Then the export process will basically organise the structure and rename folders accordingly to those event names. 

Here are some extensions to this which might cover more usecases : 
- At initialisation of Photoprism, choose a structure for the folder, this might help the photoprism folders to reflect the user current organisations.
  - YYYY/MM/DD
  - YYYY/MM
  - YYYY-MM/DD
  - etc. 
- During import process in place, use the name of the folder to create events.
- During the 'export' process, let the user choose a way to organize the folder
  - By date (no change required)
  - By date with events name on it
  - By choosing specific albums to use for the export structure (for people which are relying firstly on theme instead of events for their album)
    - Require to do lot of movement in the files (less safe for the library ?)
    - What if a photos is in more than one albums ?
    - Photos not in any chosen Albums -> use date structure for those.
-  For each photo, add all the album names as a metadata/tags (on option) 
  - Probably only during an export process as it will fully redondent with Photoprism database.

I hope this ideas of Import/Export process might help as I tried to keep Photoprism specific issues (especially Unicity of each file in the database which are imho the key) and user files management in mind.

**Comment by lastzero** [comment]
@FrouxBY There's already an event table and a page, but no business logic implemented yet. Events will be like "smart albums" based on date plus they have an optional location like for music festivals or when you travel.

**Comment by friesoft** [comment]
One major advantage for us having all changes to the filesystem reflected in photostation is you aren't locked in to the system in use. If you want to do some editing just fire up your favorite tool, no worrying if the file is renamed, cropped, ... It will just be in sync in photostation.
I guess this is my major feature I would need - keeping changes done on filesystem and in photoprism in sync. No manual thinking about Export/import/structure, everthing is Just there where you've dropped it on filesystem.
If this is nothing which is planned/desired for photoprism then I guess photoprism isn't for me.

Having smart albums is something which should be added on top using a Database. So If I create a smart Album it is only visible in photoprism but I can freely put all Photos together No Matter where they are from.

I didn't expect this discussion to skyrocket like this - so sorry for taking this issue offtopic. Is there some Community Forum or something for such discussions?

**Comment by lastzero** [comment]
@friesoft Sure, there is a community chat: https://gitter.im/browseyourlife/community

Clarification: Your originals won't be moved in ANY CASE. You can edit them and move them. Files are ONLY moved / copied if you IMPORT them from the import directory or via web upload.

If you don't use the import, duplicates in the originals directory are your own risk and a "pro" version might provide tools to deal with them if needed. Also we can use directory names or any other attributes to automatically create albums. This is not a technical issue although it might make the software more complex for users who don't need this. So at least we need to talk about the trade-offs and the amount of work we need to invest before users can enjoy the feature.

**Comment by lastzero** [comment]
During development, I delete my database all the time and then re-index the files in the originals directory. Edited metadata would get lost like this (there is always the static Exif data in the files itself), so we are going to implement XMP support. This format is supported for metadata by most photo tools. Not sure about Photostation, but Lightroom supports it. So we are even compatible with that.

**Comment by friesoft** [comment]
For photostation the metadata don't really matter - they are indexed and can be displayed but I don't think it ever modifies them. I guess photostation is best described as web file Explorer with cached photo previews and some additional Features built in top like smart albums

**Comment by lastzero** [comment]
@matthijsvdr @friesoft I've just pushed some changes that disable features who may modify the "originals" directory, namely import, upload and the RAW to JPEG converter.

Downside - obviously - is that you can not upload and also can not view / convert RAW images as JPEGs that belong to a RAW should be stored next to that RAW file. Otherwise we'd have to save the full path to images in the database and not just the path relative to originals plus you can not easily find JPEGs as they would be somewhere completely different than the rest of your files. Also you won't be able to create XMP sidecar files for you photos and obviously existing XMP files can't be updated when metadata changes, so you can not sync with Adobe Lightroom or whatever tool you are using. If your database crashes, metadata may be lost.

Also we index directory names so it will be easy to create albums or filters from that.

What do you think? Is this what you had in mind?

**Comment by matthijsvdr** [comment]
Sorry for the somewhat late reply I have been out :)

In a week or two I have some time, and I plan to setup photoprism and I'll try it out and have a better think about this so I can actually give you a good answer instead of doing it quickly now without giving much thought.

**Comment by scotws** [comment]
As somebody who just stumbled over this project while looking for a Docker-based photo storage/viewing app, let me provide the the feedback that this is the **single most important feature** for me. My photo collection includes my children's baby pictures, and yes, it is backed up multiple times and running off ZFS with snapshots, but no application gets to write to anything in that dataset, because those pictures are absolutely irreplaceable. I will have to be able to mount the docker path to the pictures read-only to even consider downloading the application. 

The model here, and this is why I don't understand why other applications don't offer this, is the video storage software Emby (probably Plex and Jellyfish as well, I just happen to know Emby). It has some strange criteria for organizing the folders and naming the media, but is perfectly fine with read-access only.

**Comment by lastzero** [comment]
@scotws No problem, read-only mode is implemented. But we already had disappointed users who were expecting us to create JPEGs / Thumbnails from their RAW files. Since we put that JPEG next to the RAW file in the same directory, it does not work with read-only. Also you can not delete files to save storage, but that could be disabled separately if needed. We would never touch any original JPEG or RAW files, never.

**Comment by lastzero** [comment]
I'll close this since read only mode is implemented in general. Please create a new issue if you feel like there is more work needed.
